Example for docs (#6)
apple/swift-collections-benchmark: Documentation/01 Getting Started.md

### Defining Benchmarks

The `CollectionsBenchmark` library makes it easy to add an executable target for running, collecting, visualizing, and comparing benchmarks from the command line.
To create a benchmark CLI tool, create a main.swift file and within it:

1. Import `CollectionsBenchmark`.

> 2. Create a `Benchmark` instance with an appropriate title.

Reviewer: What's the point of adding these steps? As far as I can tell, they are simply duplicating the code below. I'd rather we incorporated any clarifications that are needed into the comments below.

Author: I thought it made the process of adding a benchmark to a new CLI target a bit more clear. General theory was "explain what you're about to see" and then "see the work in practice". If you think it's needlessly redundant, I'll strip it out.

Reviewer: I think this is slicing it a little to fine -- "Import `CollectionsBenchmark`" looks far better as code to me. (If people are anything like me, then they'll just copy and paste the sample code and adapt it to their liking anyway. 😉)

(Aside: We badly need a chapter two of this about writing benchmark tasks -- `addSimple` only covers a small fraction of benchmarks, and the real `add` is quite a bit more subtle.)

Author: Heh - I hadn't quite come up to speed with your tooling to be able to draft anything on _that_ quickly, but yeah - I can see where it'd be a super-interesting topic, and given the complexity of it, worthy of some notes.

Author: Stripped out the step-markers and collapsed that back down to just a couple of sentences.

benchmark.main()
```

> The benchmark definition above has a title, an input type specification, and a closure to execute it. In this case, our `kalimbaOrdered` benchmark uses an array of integers as its input (populated by randomly shuffling integers in the range `0..<size`, where `size` varies between executions). On every run, it reorders its input into kalimba order, and feeds the result to the special `blackHole` function. This function discards its input -- but it does it in a way that prevents the Swift compiler from realizing that the result isn't actually used. (Otherwise, the compiler may helpfully eliminate the entire calculation, defeating the whole purpose of running the benchmark!)

Reviewer: Unlike the rest of these edits, "simply" does serve some purpose here. What if we instead replaced this with:

```suggestion
The benchmark definition above has a title, an input type specification, and a closure to execute it. In this case, our `kalimbaOrdered` benchmark uses an array of integers as its input (populated by randomly shuffling integers in the range `0..<size`, where `size` varies between executions). On every run, it reorders its input into kalimba order, and feeds the result to the special `blackHole` function. This function does nothing, discarding its input -- but it does this in a way that prevents the Swift compiler from realizing that the result isn't actually used. (Otherwise, the compiler may helpfully eliminate the entire calculation, defeating the entire purpose of running the benchmark!)
```

(If you are using Linux, you may need to replace `png` with `svg` -- you'll get the same charts, only in a different format!)

> Here is how the resulting graph looks with the results collected:

Reviewer: "Just" means "very recently" here.

```suggestion
Here is how the resulting graph looks with the results we've just collected:
```

Well that still looks like a great optimization to me -- `Array` is slightly faster at small sizes, but as the data grows, it cannot keep up with `Deque` at all. (By the way, I'm pretty sure that initial 10% difference will be reduced as the Collections package improves! `Array` has already received more than half a decade's worth of performance work, but work on `Deque` is just getting started. 😉)

The `results compare` command gets even more powerful once we have more than one benchmark. It is invaluable for getting us an objective (dare I say, ruthless) measure of how a potential change affects performance for our collection implementations.

### Benchmark Libraries

If you're anything like me, once you get the hang of writing benchmarks, it's difficult to stop. For example, the `Collections` package has several hundred (and counting!) benchmark definitions, each capturing a different aspect of data structure behavior.

With so many benchmarks, we need a way to organize them into a series of thematic charts that make sense -- trying to render them all on a single chart makes for interesting glitch art, but it isn't very practical:

![chaos.png](Assets/chaos.png)

> Here is where benchmark libraries come in! They allow us to maintain a collection of interesting charts, organized into hierarchical sections, with nice descriptive titles. Once we have a library, we can re-run the specific benchmarks it contains, and render the result in a nice document.

Reviewer: Hm, the whole point here is that rerunning the benchmarks is easier than it would be if we were doing it by hand. Losing the adjective would undersell the feature.

```suggestion
Here is where benchmark libraries come in! They allow us to maintain a collection of interesting charts, organized into hierarchical sections, with nice descriptive titles. Once we have defined a library, we can easily re-run the specific benchmarks it contains, and render the result in a nice document. We don't need to keep track of what combinations of tasks make the most useful charts, or to remember to run every task we're going to need -- the library takes care of administering these details.
```
